Recursive Bundles Proposal

pivotaljohn opened the pull request.

hackmd-deploy committed:
First pushed to GitHub

jtigger committed:
Inception.

hackmd-deploy committed:
Add examples and split the document

Create a different structure of the document
Add examples based on real bundles that will be provided with the repository

hackmd-deploy committed:
First draft of recursive bundles

This should be considered the 0.0.1 version of this proposal.
It should be in a first version that can be read and start to review by Maintainers

joaopapereira committed:
Add some examples of recursive bundles

These folders are the ones that were used to generate the proposal

cppforlife commented on imgpkg/002-recursive-bundles/README.md:
````diff
@@ -0,0 +1,280 @@
+```
+$ tree -a recursive-bundle
+.
+├── .bundles
````
i think it might be better to use `.imgpkg/bundles` instead of `.bundles` to:
- indicate that this is part of imgpkg functionality
- reduce out interface from fs perspective

joaopapereira replied:
>indicate that this is part of imgpkg functionality

I agree that this is definitely internal functionality, nevertheless, this is something that the deployer will have to be aware of.

Let's imagine you are trying to deploy an application with a recursive bundle. You will have to deploy the configuration in the root of the bundle and also go into `.imgpkg/bundles/*` in order to deploy these other bundles configuration. In the end, we would be hiding something that the deployer would have to use.

Through these lenses, I would prefer to extract the contents of each bundle to the root folder, but unfortunately, that would create the possibility of collisions of file/folder names.
I feel like the compromise of pushing these folders into `.bundles` gives the deployer a clear signal that there is something else that needs to be installed.

cppforlife replied:
> I agree that this is definitely internal functionality

i did not mean that functionality is internal. i meant that it's very much user facing but belongs to imgpkg as opposed to user included files in the bundle.

> In the end, we would be hiding something that the deployer would have to use.

deployer needs to know about images.yml. i see it as opposite to hiding -- placing it in the expected location (location that is related to imgpkg)

> gives the deployer a clear signal that there is something else that needs to be installed.

the way im seeing it is that if user of bundle requested to download referenced bundles then they clearly know to go and use them. if they did not request is as part of pull command then there is nothing for them to do. now that i think about it i didnt see any new flag to indicate to pull referenced bundles...

joaopapereira replied:
The proposal was updated to reflect this suggestion

cppforlife commented on imgpkg/002-recursive-bundles/README.md:
```diff
@@ -0,0 +1,280 @@
+
+###### Cyclic nesting
+
+To ensure that there is not cyclic nesting in the disk `imgpkg` will flatten the bundle structure to 1 level.
```
❤️

ewrenn8 replied:
Does the loss of this structure potentially create a more confusing UX for users? Say we have a high level bundle that contains multiple offerings (sub bundles), each with their own apps (installable bundles). If I want to install one of those apps, it may be nice to know which product offering, aka parent bundle, the app was from, especially if there are app name collisions. As an alternative, we could validate on push and error if there are any cycles in the bundles, which would also give the potentially helpful guarantee  that all imgpkg created bundles don't have cycles. I am not strongly opinionated here, just a thought.

joaopapereira replied:
Yes, this is definitely a hard part of the current implementation, and, it makes the UX a little hard. 
If we did prevent cyclic nesting we might be blocking an aggregator from using `imgpkg`. An example would be 2 applications, each on separate bundles, use the Rabbit bundle. If you want to create a bundle with these 2 applications you cannot, or you will have to break the individual application bundles and create your own bundle.
This can happen with Cloud Foundry or some other big system that contains multiple single-purpose applications that are bundled separately but you also want to provide a full installation.

To address some of these concerns I think we could eventually enhance this UX by allowing the bundle creators to provide an annotation that would tell `imgpkg` the name of the folder for a particular bundle in a later proposal. Do you think this would be helpful?

Note: Eventually if we were to provide such annotation we would have to ensure that there are no cyclical nesting.

joaopapereira replied:
@timothysc is your comment about this ^
If so I was talking about a Rabbit Bundle that would contain Rabbit OCI Image + default configurations. 
Imagine this as the Bundle that the RabbitMQ team could provide for you to install in your cluster.

ewrenn8 replied:
Ah, yea, I guess if we preserved structure, since it is going on to a filesystem, each bundle would be limited to one parent, so we would have to write the bundle multiple times if it appeared multiple times, which seems less than ideal.

timothysc replied:
So long as the lock file is clear I'm fine with flatting so long as users can nest to their hearts content provided there are 0 cycles. It needs to be a DAG.

joaopapereira replied:
There will be no limit to the depth you can go with nested bundles. The only caveat is the more you add to a bundle the more time it will take to get everything.

The image lock file for each bundle will contain all the information that you would need to find the folder that will contain the bundle configuration you are looking for. Maybe not the location of the bundle but at least the information on how might you find it, like the SHA for the currently proposed placement in the disk.

joaopapereira replied:
I believe the concerns in this thread were addressed in newer reviews of the proposal

cppforlife commented on imgpkg/002-recursive-bundles/examples/bundle-1/.imgpkg/images.yml:
```diff
@@ -0,0 +1,7 @@
+- image: ghcr.io/k14s/design-docs/simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
+  annotations:
+    kbld.carvel.dev/id: my.registry.io/simple-application
+- image: ghcr.io/k14s/design-docs/utilities@sha256:47ae428a887c41ba0aedf87d560eb305a8aa522ffb80ac1c96a37b16df038e0f
```
we would need to make sure that kbld is happy to skip images that arent annotated with `kbld.carvel.dev/id`. not sure if that's already being done?

joaopapereira replied:
I just tried it and it does not look happy

```
$ cat tmp/images1.yml
---
apiVersion: imgpkg.carvel.dev/v1alpha1
images:
- annotations:
  image: gcr.io/cf-k8s-lifecycle-tooling-klt/demo-projects-app@sha256:27c54fe47e8b86576557100a4a0e0ee7c358629f40c6a7c93940f656aba7bcd2
kind: ImagesLock
```

```
$ kbld -f deployment/config/app.yml -f tmp/images1.yml
kbld: Error: Validating imageslock/ (imgpkg.carvel.dev/v1alpha1) cluster:
  Validating Overrides[0]:
    Expected Image or ImageRepo to be non-empty
```

cppforlife commented on imgpkg/002-recursive-bundles/README.md:
````diff
@@ -0,0 +1,280 @@
+
+```
+Warning: .bundles folder could not be added to the bundle.
+         This folder is used to store configurations of recursive bundles and cannot be checked into the bundle
````
What do you think of shorter warning like `Warning: Ignoring ".imgpkg/bundles" directory (directory is only used for keeping pulled dependent bundles)`?

hackmd-deploy committed:
Add Image listing feature

joaopapereira committed:
Shorten the warning message for `.bundles`

danielhelfand commented on imgpkg/002-recursive-bundles/README.md:
```diff
@@ -0,0 +1,304 @@
+
+## Pushing a bundle
+
+The major proposed change in this section is to remove the validation done today that checks if the bundle being pushed contains bundles in the `.imgpkg/images.yml` Lock file.
```
Has there been any thought about keeping `images.yml` exclusive to images and using something like the [BundleLock](https://carvel.dev/imgpkg/docs/latest/resources/#bundlelock-configuration) in the case of nested bundles?

joaopapereira replied:
I think this separation could be an interesting idea, nevertheless, having a single lock file with all the images feels more natural. This separation would help `imgpkg` implementation because now it knows which image is based on the file it picks it up from. But it would make the life of the bundle create harder because this separation manually would have to be done and maintained by hand.
Also, it would increase the complexity of the image validation because now we would have to ensure that the creator placed bundles in one Lock file and images in a different Lock File.

After going through these pros and cons I would prefer to keep all the images in a single Lock file, to give the user the perception that `imgpkg` handles images. These images could be bundles or regular images, but in the end, they are images.

danielhelfand replied:
Yeah, I am not strongly opinionated on it and agree it feels like better UX to have a single file. I think it calls into question the need for a BundleLock file if this isn't a use case for it. But that's outside the scope of this.

cppforlife replied:
+1 on single file for all referenced assets.

ewrenn8 commented on imgpkg/002-recursive-bundles/README.md:
````diff
@@ -0,0 +1,304 @@
+7 directories, 9 files
+```
+
+The folder name will be the sha256-{SHA} where SHA is the SHA256 of the bundle OCI Image. The mapping between the folder names and the origin images can be found in the `.imgpkg/images.yml` of the bundle that included this bundle
````
Why dont we just put the image name + sha (maybe even just image@sha256... from the lockfile) as the filename instead of forcing users to manually decode which digest is which application bundle? The digest in the filename will ensure there are no collisions, and, if there are, we deduped our pull wrong.

joaopapereira replied:
I have some concerns about the image name + sha approach.
- What do we consider image name? Only the repository or the full image name with the registry?
  - If we use only the repository name we might have to do some decoding by hand because we might have multiple images in the same repository
  - If we have the full image with registry this will become a pretty long name and we would also have problems with the sub-folder structure that we would have to convert the / -> _ or something of the sorts


The intention behind the `imgpkg info -b` is to also help with this particular problem where you could easily look at the image name and its sha so you could do the translation.

ewrenn8 replied:
Yea, thats true. I guess since the users will no longer have the context of where a certain bundle was referenced from, it is a bit harder to distinguish what the name means, i.e. knowing the parent bundle could helps distinguish overlapping names. Though, I am not sure how likely colliding bundle names are, so maybe defaulting to putting the repo name there and then qualifying it with some info about the registry in the case of a collision would help?

It just feels like a poor experience to have to find a specific digest in a list of potentially hundreds of images in order to find the app I'm looking for on disk.

Also, will the info command recursively compile a list of ALL images/bundles referenced by a bundle, and, if so, do we have an idea for the scale we expect? Will it also be able to provide me info from an on disk bundle? If the scale is huge and I just spent a significant amount of time pulling everything, I probably wouldn't want to wait for imgpkg to essentially pull everything again just to create the list of images for the info command, especially if I need to do that every time I'm trying to find a new app on disk and didn't have the foresight to save the output of info.

joaopapereira replied:
> Also, will the info command recursively compile a list of ALL images/bundles referenced by a bundle, and, if so, do we have an idea for the scale we expect? Will it also be able to provide me info from an on disk bundle? If the scale is huge and I just spent a significant amount of time pulling everything, I probably wouldn't want to wait for imgpkg to essentially pull everything again just to create the list of images for the info command, especially if I need to do that every time I'm trying to find a new app on disk and didn't have the foresight to save the output of info.

Yes, the idea would be for the info command to provide information about all nested bundles.
I like the idea of the info command to read from disk, or even eventually we could think about the option of creating some sort of cache in disk with information on the images. But I will leave this idea for a later proposal.

> It just feels like a poor experience to have to find a specific digest in a list of potentially hundreds of images in order to find the app I'm looking for on disk.

Do you think that if the info command reads from disk this experience would be improved?


Action Item for me: Add the option for the info to read from disk instead.

joaopapereira replied:
Just updated the proposal with 2 changes:
1. Add the command to read the information about a bundle that was previously pulled to disk
2. Add verbiage around the goals of the document and point out that in this proposal the changes are oriented to Applications used to interact with the bundle. Eventually, in a future proposal, we can create a better visualization of the bundles on disk to help with some of these UX problems.

ewrenn8 replied:
Reading an on disk bundle may not be useful since we prefer rewritting the imageslock if all the images are present in the bundle repo. This means users will most likely end up with the bundle repo name + digest, which isn't very informative since image names are lost there.

This may point back to the idea that had been proposed of somehow splitting the `.imgpkg` dir in to its own layer so that we can pull that easily without needing everything else in the image.

joaopapereira replied:
> Reading an on disk bundle may not be useful since we prefer rewritting the imageslock if all the images are present in the bundle repo. This means users will most likely end up with the bundle repo name + digest, which isn't very informative since image names are lost there.

Yes, indeed you are right, that last change to the proposal might not be very useful after all. Going to revert that change.

Regarding the layering, I like the idea and we should create a proposal to define that.

joaopapereira replied:
I believe this issue has been addressed in the newer review of the proposal

joaopapereira committed:
Change pull behavior

After some conversations about pull the following change was made:
- `imgpkg pull` command will no longer by default download all the nested bundles
- New flag `imgpkg pull -r` will download all the nested bundles

we decided to opt into this approach because if a bundle contains a high amount of nested bundles or nested bundles that the user might not need, it is better to do not to download everything unless the user explicitly requests it.

timothysc reviewed (commented):
How far down does the rabbit how go.  This example is only 1 layer deep.

joshrosso commented on imgpkg/002-recursive-bundles/README.md:
```diff
@@ -0,0 +1,328 @@
+
+When using the `pull` command without the `-r` flag the behavior is not changed. Nothing will be dowloaded from the nested bundles.
+
+When downloading a bundle that contains other bundles to disk using the `pull -r` command will work as currently, except for that it will download the nested bundles into a hidden folder called `.bundles`.
```
As a user, I think i'd be surprised that the recursive behavior is not the default.

And that I'd need to know if a bundle references other bundles, then add the `-r` flag.

joaopapereira replied:
That in fact was the original proposal, but we updated it because we thought that it would give us a little more flexibility for possible future features.
If we think of a bundle to be a group of images and a configuration for an application the default behavior should be pulled recursively. But if we think about a bundle as a repository of applications you might not want to recursively pull all the applications to read this repository information.

To be fair we do not know yet what will be the more common use case for the recursive bundles. Given the information we have today we believe that making the default behavior not pull recursively is going to allow us to cater to possible future features more easily.

timothysc replied:
I agree with Josh fwiw, I would consider non-recursive to the be exception vs. the norm.  
+1 to be recursive by default.

joaopapereira replied:
Given the current use case we have for Recursive Bundles (bundle as a repository of applications) we believe that making the default behavior not pull recursively would be more helpful for the users of `imgpkg`.
In the future, if we see a change in the usage of `imgpkg` we might change this behavior default behavior.

joaopapereira committed:
Add language around cyclic nesting

After some back and forth the proposal was updated to include some Goals to better address the intent behind this approach.

This first iteration of the feature will cater for applications to be used in conjunction with `imgpkg` that know how to read the output format provided. In a later iteration, a better Human UX might be added to help visualize the structure that each bundle in the disk has.

joaopapereira committed:
Read bundle information from disk

Add a new scenario to ensure that if the user pre-pulled the bundle, `imgpkg` can read the contents from the file system instead of having to retrieve all the information from the OCI registry

joaopapereira committed:
Revert "Read bundle information from disk"

This reverts commit c941e5add0b85c639abd8a24db37473ae9169bf8.

joaopapereira committed:
Change the proposal status

joaopapereira committed:
Move the bundles pull folder

Changed the location where the bundles will be downloaded too. Instead of being downloaded to the root of the folder inside `.bundles` the recursive bundles will be downloaded to `.imgpkg/.bundles` this will minimize the external footprint of `imgpkg`

joaopapereira committed:
Unhide the bundles folder inside .imgpkg
